### ISSUE TITLE
Compile errors with a generic base class and class mixins using declarative typings

### ISSUE THREAD

**Issue by agubler** [Microsoft/TypeScript#14017]
**TypeScript Version:**  2.2.0 (rc) & 2.2.0-dev.20170211

I am having trouble creating consumable typings when using class mixins for a generic class from a core library - github example repro here https://github.com/agubler/typescript-mixin-typings.

Consider the following provided by a core library

```ts
// BaseClass.ts
export type Constructor<T> = new (...args: any[]) => T;

export class MyBaseClass<T> {
    baseProperty: string;

    constructor(value: T) {}
}
```

```ts
// MixinClass.ts
import { Constructor, MyBaseClass } from './BaseClass.ts';

export interface MyMixin {
    mixinProperty: string;
}

export function MyMixin<T extends Constructor<MyBaseClass<any>>>(base: T): T & Constructor<MyMixin> {
    mixinProperty: string;
}
```

```ts
// FinalClass
export class MyExtendedClass extends MyMixin(MyBaseClass)<string> {
	extendedClassProperty: number;
}
```

compiling with `declarations` `true`

```json
{
		"declaration": true,
		"emitDecoratorMetadata": true,
		"experimentalDecorators": true,
		"module": "umd",
		"moduleResolution": "node",
		"noImplicitAny": true,
		"noImplicitThis": true,
		"removeComments": false,
		"sourceMap": true,
		"strictNullChecks": true,
		"target": "es5"
}
```

Usage in downstream consumer project:

```ts
import { MyExtendedClass } from 'core-library/FinalClass';
import { MyMixin } from 'core-library/MixinClass';

// constructor signature lost
const myExtendedClass = new MyExtendedClass('string');

// MyExtendedClass no longer type MyBaseClass
const AnotherMixedClass = MyMixin(MyExtendedClass);
```

Errors:
`5 col 25 error| Supplied parameters do not match any signature of call target. [typescript/tsuquyomi]`

`8 col 35 error| Argument of type 'typeof MyExtendedClass' is not assignable to parameter of type 'Constructor<MyBaseClass<any>>'. Type 'MyExtendedClass' is not assignable to type 'MyBaseClass<any>'. Property 'baseProperty' is missing in type 'MyExtendedClass'. [typescript/tsuquyomi]`

**Comment by ahejlsberg** [comment]
Your example has several syntax errors and missing imports in it. However, once I fix it up the following compiles with no errors:

```ts
// BaseClass.ts
export type Constructor<T> = new (...args: any[]) => T;

export class MyBaseClass<T> {
    baseProperty: string;
    constructor(value: T) {}
}
```
```ts
// MixinClass.ts
import { Constructor, MyBaseClass } from './BaseClass';

export interface MyMixin {
    mixinProperty: string;
}

export function MyMixin<T extends Constructor<MyBaseClass<any>>>(base: T): T & Constructor<MyMixin> {
    return class extends base {
        mixinProperty: string;
    }
}
```
```ts
// FinalClass.ts
import { MyBaseClass } from './BaseClass';
import { MyMixin } from './MixinClass';

export class MyExtendedClass extends MyMixin(MyBaseClass)<string> {
    extendedClassProperty: number;
}
```
```ts
// Main.ts
import { MyExtendedClass } from './FinalClass';
import { MyMixin } from './MixinClass';

const myExtendedClass = new MyExtendedClass('string');

const AnotherMixedClass = MyMixin(MyExtendedClass);
```

**Comment by agubler** [comment]
@ahejlsberg Thanks for looking was it the GitHub repro that was wrong or my bad on the code on the issue

**Comment by agubler** [comment]
The honest answer is I want you to be correct, but I really struggled to get anything compiling given the typings generated from my example

**Comment by agubler** [comment]
@ahejlsberg did you try `main.ts` compiling from the outputted typings rather than the typescript files in the same project?

**Comment by ahejlsberg** [comment]
Ah, I see what the problem is now. The declaration file emitter is emitting incorrect code in FinalClass.d.ts:

```ts
// FinalClass.d.ts
import { MyBaseClass } from './BaseClass';
import { MyMixin } from './MixinClass';
export declare class MyExtendedClass extends MyBaseClass<string> & MyMixin {
    extendedClassProperty: number;
}
```

The `extends` clause of a class needs to specify an expression, but the declaration file emitter emits a type. It should instead report an error if the expression isn't a simple variable.

You can work around the issue by introducing a `const` that represents the mixin base class:

```ts
import { Constructor, MyBaseClass } from './BaseClass';
import { MyMixin } from './MixinClass';

export const MyMixedBaseClass = MyMixin(MyBaseClass);

export class MyExtendedClass extends MyMixedBaseClass<string> {
	extendedClassProperty: number;
}
```

That produces the following correct output:

```ts
import { Constructor, MyBaseClass } from './BaseClass';
import { MyMixin } from './MixinClass';
export declare const MyMixedBaseClass: typeof MyBaseClass & Constructor<MyMixin>;
export declare class MyExtendedClass extends MyMixedBaseClass<string> {
    extendedClassProperty: number;
}
```

**Comment by agubler** [comment]
Great, thank you - I will try that out now (perhaps I should have just put the emitted code from `FinalClass.d.s` in the issue!).

Is it just a side effect that is works when a generic is not passed? Because without a generic the emitted code is 

```ts
export declare class MyExtendedClass extends MyBaseClass & MyMixin {
    extendedClassProperty: number;
}
```

So still type but behaves as expected?

**Comment by sandersn** [comment]
@agubler extending an intersection still doesn't work though &mdash; it fails in the parser (and has no supporting code in the checker either). After some discussion, we're going to make it a d.ts error for 2.2 and consider supporting `&` in `extends` clauses in 2.3.

**Comment by agubler** [comment]
@sandersn Thanks for the feedback, the work around to assign the result of the mixins to a intermediary `const` is still fine though? As per @ahejlsberg's suggestion?

**Comment by sandersn** [comment]
Yes, it's just a limitation of the parsing and checking of heritage clauses &mdash; they expect a type reference, not an arbitrary expression.

**Comment by agubler** [comment]
I thought as much, at least with it being an error the issue will be highlighted before we try to consume the published module(s).

**Comment by sandersn** [comment]
Initial fix is up at #14074

I'll look at actually parsing intersections after this so that it's no longer an error.
